feat(spanner): add GCP standard otel attributes

rahul2393 opened the pull request:
**Key Changes**:
- **Centralized Span Management**:
   - Introduced trace.go with startSpan/endSpan helpers to standardize span lifecycle management, including error recording and status propagation.
   - All spans are now uniformly tagged with common attributes for diagnostics (gcp.client.version, gcp.client.repo, gcp.client.artifact) and resource context (db.name, instance.name, cloud.region).
- **Refactored and Enriched Span Attributes**:
  - Added a generic setSpanAttributes function in grpc_client.go to dynamically set request-specific attributes (db.statement, transaction.tag, etc.). The function also omits attributes with empty values.
  - Added trace attributes to capture server-timing headers on streaming and unary RPCs.
- **Consistent Naming**:
  - All span names are now prefixed with cloud.google.com/go/spanner. for uniform identification in tracing backends.

sakthivelmanii commented on spanner/trace.go:
```diff
@@ -0,0 +1,74 @@
+
+// getCommonTraceOptions includes the common attributes used for Cloud Trace adoption tracking.
+func getCommonTraceOptions() []trace.SpanStartOption {
+	opts := []trace.SpanStartOption{
```
do we need to assign to a variable?

sakthivelmanii commented on spanner/trace.go:
```diff
@@ -0,0 +1,74 @@
+func endSpan(ctx context.Context, err error) {
+	span := trace.SpanFromContext(ctx)
+	if err != nil {
+		span.SetStatus(otelcodes.Error, err.Error())
```
Don't we need this implementation? https://github.com/googleapis/google-cloud-go/blob/main/internal/trace/trace.go#L57

rahul2393 commented:
<img width="1912" height="670" alt="Screenshot 2025-08-02 at 3 56 38 AM" src="https://github.com/user-attachments/assets/b38c0a2c-ae65-40b1-8a4f-f24e322667a7" />

olavloite commented on spanner/grpc_client.go:
```diff
@@ -171,7 +174,64 @@ func (g *grpcSpannerClient) DeleteSession(ctx context.Context, req *spannerpb.De
+	}
+	for t, v := range latencyMap {
+		span.SetAttributes(
+			attribute.Float64(t[:3]+".latency_ms", float64(v.Nanoseconds())/1e6),
```
nit: this could use a small comment explaining what `t[:3]` does (gfe or afe latency). And are we absolutely sure that this will never change in the future? What if we introduce some new header key in the future that is shorter, and someone is still using this version of the client? Maybe we should only do the `t[:3]` for the well-known header keys, and include the full header key for other values (or completely skip them).

olavloite reviewed (commented):
Could we also modify some of the existing test cases and/or add new test cases that verify that the addtional span attributes are present? The main reason for wanting that is not because I think that they are not there, but to prevent accidental regressions in the future where someone by accident removes them, or changes the key of (some of) them.

rahul2393 commented on spanner/client.go:
```diff
@@ -718,8 +720,10 @@ func metricsInterceptor() grpc.UnaryClientInterceptor {
-		metrics := parseServerTimingHeader(md)
-		mt.currOp.currAttempt.setServerTimingMetrics(metrics)
+		latencies := parseServerTimingHeader(md)
+		span := otrace.SpanFromContext(ctx)
```
We anyways need parseServerTimingHeader because of native metrics, in this PR we just pass it to setGFEAndAFESpanAttributes and if customer is not using tracing it won't be captured et-all
```
if !span.IsRecording() {
		return
}
```
And yes it always returns a useful value both for cloud path and direct path

rahul2393 commented on spanner/grpc_client.go:
```diff
@@ -171,7 +174,64 @@ func (g *grpcSpannerClient) DeleteSession(ctx context.Context, req *spannerpb.De
+
+	var attrs []attribute.KeyValue
+
+	if r, ok := any(req).(interface {
```
I added a benchmark test `BenchmarkSetSpanAttributes`


### High-Level Summary

1.  **Not Recording is Cheap:** When the span is not recording, the function's overhead is negligible and constant (~64 ns, 0 allocations). The function correctly returns early, making it very efficient in this state.
2.  **Recording Cost is Proportional to Work:** When the span *is* recording, the performance cost (latency and allocations) is directly proportional to the number and type of attributes being set on the span.
3.  **Tail Latency is High:** The p99 latencies are significantly higher than the median (p50) latencies. This indicates that while the function is fast on average, there are occasional spikes where it can take much longer (e.g., 2.6µs vs a median of 166ns for `ExecuteSqlRequest`). This is common in tracing systems and can be caused by factors like garbage collection, context switching or internal locking within the tracing SDK.

---

### Detailed Breakdown

#### 1. Not-Recording Spans

Across all request types, the performance is consistently fast and predictable:
*   **Latency:** ~64 ns
*   **Allocations:** 0 allocs/op (the 8 bytes reported are likely from the benchmark harness itself).
*   **p50/p90/p99:** All hover around 41-42 ns.

This confirms that the `if !span.IsRecording()` check is effective at making the function call a cheap no-op when tracing is disabled or sampled out.

#### 2. Recording Spans

Here, performance varies based on the request type, as expected.

*   **`ExecuteBatchDmlRequest` (Worst Case):**
    *   **Latency:** ~380 ns (p50: 250 ns, p99: 2.75 µs)
    *   **Allocations:** 6 allocs/op (708 B)
    *   **Reason:** This is the most expensive case because the function iterates over a slice of DML statements, creates a new slice of strings (`sqls`), and then adds multiple attributes. This involves more logic and memory allocation than other types.

*   **`ExecuteSqlRequest` & `ReadRequest`:**
    *   **Latency:** ~270-280 ns (p50: 166 ns, p99: ~2.6 µs)
    *   **Allocations:** 3 allocs/op (612 B)
    *   **Reason:** These requests set three attributes (`transaction.tag`, `statement.tag`, and `db.statement` or `db.table`), leading to moderate latency and allocations.

*   **`CommitRequest`:**
    *   **Latency:** ~200 ns (p50: 125 ns, p99: ~2.0 µs)
    *   **Allocations:** 2 allocs/op (302 B)
    *   **Reason:** This is faster as it only sets two attributes from the request options.

*   **`PartitionQueryRequest` & `PartitionReadRequest`:**
    *   **Latency:** ~130 ns (p50: 83 ns, p99: ~0.4 µs)
    *   **Allocations:** 1 alloc/op (123 B)
    *   **Reason:** You correctly identified that these requests don't always have `RequestOptions`. After your change removing them, these requests only set one attribute (`db.statement` or `db.table`), making them much faster and more memory-efficient than the other requests.

*   **Requests with No Attributes (`BatchWriteRequest`, `Empty struct`):**
    *   **Latency:** ~80 ns
    *   **Allocations:** 0 allocs/op
    *   **Reason:** The performance is very close to the `not-recording` case. The slightly higher latency comes from the additional type assertion checks inside the function which ultimately find no attributes to set.

### Conclusion

The `setSpanAttributes` function is well-optimized. The primary performance cost is incurred only when a span is actively recording, and this cost scales linearly with the complexity of the request object. The worst-case median latency is low (~250 ns for a batch DML request), though tail latencies can be an order of magnitude higher. For requests that don't set any attributes, the overhead is minimal.

rahul2393 commented:
@olavloite I updated test `TestSpannerTracesWithOpenTelemetry` to validate newly added attributes.
I also added benchmark tests to check how much `setSpanAttributes` overhead has, I added the summary [here](https://github.com/googleapis/google-cloud-go/pull/11652#discussion_r2249558430)
